Add native GPU support. (#504)
JuliaMath/Interpolations.jl: docs/src/devdocs.md

General `AbstractArray`s may be indexed with `WeightedIndex` indices,
and the result produces the interpolated value. In other words, the end result
is just `itp.coefs[wis...]`, where `wis` is a tuple of `WeightedIndex` indices.

> To make sure this overloading is effective, we wrapper the `coefs` with `InterpGetindex`，

Reviewer: ```suggestion
To make sure this overloading is effective, we wrap the `coefs` with `InterpGetindex`，
```

It makes good use of *tuple* manipulations, sometimes called "lispy tuple programming."
You can search Julia's discourse forum for more tips about how to program this way.
It could alternatively be done using generated functions, but this would increase compile time considerably and can lead to world-age problems.

### GPU Support

> At present, `Interpolations.jl` supports interpolant usage on GPU via broadcast mechanism.

Reviewer: ```suggestion
At present, `Interpolations.jl` supports interpolant usage on GPU via broadcasting.
```

A basic work flow looks like:
```julia
julia> using Interpolations, Adapt, CUDA # Or any other GPU package

julia > itp = Interpolations.interpolate([1, 2, 3], (BSpline(Linear()))); # construct the interpolant object on CPU

julia> cuitp = adapt(CuArray{Float32}, itp); # adapt it to GPU memory

julia > cuitp.(1:0.5:2) # call interpolant object via broadcast

julia> gradient.(Ref(cuitp), 1:0.5:2)
```

> To achieve this, an `ITP <: AbstractInterpolation` should define it's own `Adapt.adapt_structure(to, itp::ITP)`, which constructs a new `ITP` with the adapted
> fields (`adapt(to, itp.fieldname)`) of `itp`. The field adaption could be skipped
> if we know that it has been GPU-compatable, e.g. a bit range.

Reviewer: ```suggestion
To achieve this, an `ITP <: AbstractInterpolation` should define its own `Adapt.adapt_structure(to, itp::ITP)`.
`adapt_structure` constructs a new `ITP` with the adapted fields of `itp` as defined by `adapt(to, itp.fieldname)`. The field adaption could be skipped if we know that it is been GPU-compatable, e.g. a bit range.
```

Also what do you mean by "bit range" here. Do you mean bit type where `isbitstype` is true?

Author: Yes, has been replaced with "a `isbit` range"

!!! note

>     Some adaptor might change the storage type. Please ensure that the adapted `itp`
>     has the corrent `eltype`.

Reviewer: ```suggestion
    Some adaptors may change the storage type. Please ensure that the adapted `itp`
    has the correct element type via the method `eltype`.
```

> Also, all GPU-compatable `AbstractInterpolation`s should define their own `Interpolations.root_storage_type`. This function help us injecting into the broadcast mechanism by overloading the default `BroadcastStyle`. See [Customizing broadcasting](https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting) for more details.

Reviewer: ```suggestion
Also, all GPU-compatable `AbstractInterpolation`s should define their own `Interpolations.root_storage_type`. This function allows us to modify the broadcast mechanism by overloading the default `BroadcastStyle`. See [Customizing broadcasting](https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting) for more details.
```
